packages/hardhat-core/src/internal/hardhat-network/provider/utils/fakeTransaction.ts

```diff
@@ -0,0 +1,103 @@
+  BN,
+  bufferToInt,
+  rlphash,
+  toBuffer,
+  unpadBuffer,
+} from "ethereumjs-util";
+
+import { TransactionParams } from "../node-types";
+
+export interface FakeTxData extends TxData {
+  from?: AddressLike;
+}
+
+type ArrayWithFrom<T> = T[] & { from?: string };
+
+export class FakeTransaction extends Transaction {
+  public static fromValuesArray(
+    values: ArrayWithFrom<Buffer>,
+    opts: TxOptions = {}
+  ) {
+    const [nonce, gasPrice, gasLimit, to, value, data, v, r, s] = values;
+
+    const emptyBuffer = Buffer.from([]);
+
+    return new FakeTransaction(
+      {
+        from: values.from,
+        nonce: new BN(nonce),
+        gasPrice: new BN(gasPrice),
+        gasLimit: new BN(gasLimit),
+        to: to !== undefined ? new Address(to) : undefined,
+        value: new BN(value),
+        data: data ?? emptyBuffer,
+        v: v !== undefined && !v.equals(emptyBuffer) ? new BN(v) : undefined,
+        r: r !== undefined && !r.equals(emptyBuffer) ? new BN(r) : undefined,
+        s: s !== undefined && !s.equals(emptyBuffer) ? new BN(s) : undefined,
+      },
+      opts
+    );
+  }
+  public fakeFrom: Address;
+
+  constructor(txParams: FakeTxData, txOptions: TxOptions) {
+    const _fakeFrom = new Address(toBuffer(txParams.from));
+
+    // `delete` requires that property is optional. Really
+    // harhdhat/TransactionParams is used as @ethereumjs/TxData
+    // (which has no `from`) here
+    delete (txParams as any).from;
+
+    super(txParams, { ...txOptions, freeze: false });
+    this.fakeFrom = _fakeFrom;
+  }
+
+  public getSenderAddress(): Address {
+    return this.fakeFrom;
+  }
+
+  // Ported from ethereumjs-tx `tx.hash(false)`
+  public hash(): Buffer {
+    let items;
+
+    if (this._implementsEIP155()) {
+      items = [
+        ...this.raw().slice(0, 6),
+        toBuffer(this.common.chainId()),
+        unpadBuffer(toBuffer(0)),
+        unpadBuffer(toBuffer(0)),
+      ];
+    } else {
+      items = this.raw().slice(0, 6);
+    }
+
+    // create hash
+    return rlphash(items);
+  }
+
+  private _implementsEIP155(): boolean {
+    const onEIP155BlockOrLater = this.common.gteHardfork("spuriousDragon");
+
+    if (!this.isSigned()) {
+      // We sign with EIP155 all unsigned transactions after spuriousDragon
+      return onEIP155BlockOrLater;
+    }
+
+    // EIP155 spec:
+    // If block.number >= 2,675,000 and v = CHAIN_ID * 2 + 35 or v = CHAIN_ID * 2 + 36, then when computing
+    // the hash of a transaction for purposes of signing or recovering, instead of hashing only the first six
+    // elements (i.e. nonce, gasprice, startgas, to, value, data), hash nine elements, with v replaced by
+    // CHAIN_ID, r = 0 and s = 0.
+    const v = bufferToInt(this.v!.toBuffer());
+
+    const vAndChainIdMeetEIP155Conditions =
+      v === this.common.chainId() * 2 + 35 ||
+      v === this.common.chainId() * 2 + 36;
+    return vAndChainIdMeetEIP155Conditions && onEIP155BlockOrLater;
+  }
+}
```

Comment:
Could use simpler version from ForkTransaction

Reply:
All removed in 36add01. FakeTransaction now uses the underlying class's: LegacyTransaction#hash.

NB: There are different `hash` implementations for Legacy and 2930 transactions:
+ [Legacy][1]
+ [2930][2]

[1]: https://github.com/ethereumjs/ethereumjs-monorepo/blob/3d9854db3ce0e15263ff4c3c0db9eb31cbb967d3/packages/tx/src/legacyTransaction.ts#L166-L168
[2]: https://github.com/ethereumjs/ethereumjs-monorepo/blob/3d9854db3ce0e15263ff4c3c0db9eb31cbb967d3/packages/tx/src/eip2930Transaction.ts#L288-L294